### PR TITLE
[DEVOPS-464] stack2nix: Pin down hackage db.

### PR DESCRIPTION
**Pull request by ksaric**
https://iohk.myjetbrains.com/youtrack/issue/DEVOPS-464

Currently hitting https://github.com/input-output-hk/stack2nix/issues/66, should be tested.

Fixes https://github.com/input-output-hk/stack2nix/issues/52

### CODE REVIEW

File: src/Stack2nix.hs [hunk 1]
```diff
@@ -91,7 +93,7 @@ stack2nix args@Args{..} = do
 
     checkRuntimeDeps :: IO ()
     checkRuntimeDeps = do
-      checkVer "cabal2nix" "2.2.1"
+      checkVer "cabal2nix" "2.7"
```
Comment (domenkozar) [inline src/Stack2nix.hs]: It's not due to this PR, but this check passes for example with cabal2nix 2.5

**Comment by domenkozar** [comment]
Works locally for me with caveats:

- cabal2nix 2.7 is not in nixpkgs yet, so it had to be built manually (this blocks this PR until we have nixpkgs with it)
- lots of debugging output ended up in stdout, so I had to strip it out
- stack2nix changed hashes:

```
       mkDerivation {
           pname = "Cabal";
           version = "2.0.0.2";
-          sha256 = "0chhl2113jbahd5gychx9rdqj1aw22h7dyj6z44871hzqxngx6bc";
+          sha256 = "6c99fe6cc71f868308f946fa76a0105c05895b4e1d32ff4a836ac91182a01
032";
           libraryHaskellDepends = [
             array
             base
```

Invoked 

```
$ PATH=/nix/store/7cbhi0yjm185cb987d0vbsnwz410179r-cabal2nix-2.7/bin:$PATH /nix/store/16zdkwsxh181ilcjmmmws8ydbssdifk8-stack2nix-0.1.3.0/bin/stack2nix --hackage-snapshot 2017-10-30T12:18:35Z . > stack2nix.nix
```

File: src/Stack2nix.hs (hunk 1 above)
Comment (jmitchell) [reply src/Stack2nix.hs]: This is fixed in the latest commit. The previous version parser was interpreting the following as `22.7` (first 2 from `cabal2nix`).

```
$ cabal2nix --version
cabal2nix 2.7
```

Now we have a regular expression and a longest match heuristic and test cases to verify a few cases, including the scenario in #67.

**Comment by jmitchell** [comment]
The latest commit switches to using cabal2nix 2.7 as a library, eliminating the need for the runtime binary dependency.

**Comment by domenkozar** [comment]
@jmitchell seems we have a bit more impurity. Since you added `happy` to stack, it didn't generate the package for it.

Also I had to apply following diff:

```
diff --git a/default.nix b/default.nix
index 6940458..4117e62 100644
--- a/default.nix
+++ b/default.nix
@@ -6,8 +6,7 @@ with pkgs.haskell.lib;
   overrides = self: super: {
     stack2nix = justStaticExecutables super.stack2nix;
 
-    # https://github.com/input-output-hk/stack2nix/issues/51
-    Cabal = pkgs.haskell.packages.ghc802.Cabal_2_0_0_2;
+    Cabal_2_0_0_2 = super.Cabal;
 
     # https://github.com/NixOS/cabal2nix/issues/146
     hinotify = if pkgs.stdenv.isDarwin then self.hfsevents else super.hinotify;
```

**Comment by domenkozar** [comment]
@jmitchell also we'll have to use threaded runtime now

**Comment by domenkozar** [comment]
Done.

**Comment by domenkozar** [comment]
@jmitchell it was confirmed this fixes https://github.com/input-output-hk/stack2nix/issues/66

File: src/Stack2nix.hs [hunk 2]
```diff
@@ -62,38 +55,18 @@ stack2nix args@Args{..} = do
...
-          unless (ver >= parseVer minVer) $ error $ unwords ["ERROR:", prog, "version must be", minVer, "or higher. Current version:", maybe "[parse failure]" showVersion ver]
-        (ExitFailure _, _, err)  -> error err
-
     checkRuntimeDeps :: IO ()
     checkRuntimeDeps = do
-      checkVer "cabal2nix" "2.2.1"
-      checkVer "git" "2"
-      checkVer "cabal" "1"
+      assertMinVer "git" "2"
+      -- assertMinVer "cabal" "2"
```
Comment (domenkozar) [inline src/Stack2nix.hs]: @jmitchell we are still using `cabal-install`, so this should be uncommented

File: src/Stack2nix/External/Stack.hs [hunk 3]
```diff
@@ -142,10 +146,11 @@ planAndGenerate boptsCli baseDir outDir remoteUri revPkgs argRev threads doAfter
   plan <- withLoadPackage $ \loadPackage ->
     constructPlan mbp baseConfigOpts locals extraToBuild localDumpPkgs loadPackage sourceMap installedMap (boptsCLIInitialBuildSteps boptsCli)
   let pkgs = prioritize $ planToPackages plan ++ revPkgs
-  -- liftIO $ hPutStrLn stderr $ "plan:\n" ++ ppShow pkgs
+  liftIO $ hPutStrLn stderr $ "plan:\n" ++ show pkgs
 
-  void $ liftIO $ mapM_ (\pkg -> cabal2nix ("cabal://" ++ pkg) Nothing Nothing (Just outDir)) $ words "hscolour stringbuilder"
-  void $ liftIO $ mapPool threads (genNixFile baseDir outDir remoteUri argRev) pkgs
+  hackageDB <- liftIO $ loadHackageDB Nothing hSnapshot
```
Comment (domenkozar) [inline src/Stack2nix/External/Stack.hs]: This now doesn't pass hackage db snapshot anymore

File: stack2nix.cabal [hunk 4]
```diff
@@ -1,5 +1,5 @@
 name:                stack2nix
-version:             0.1.3.0
+version:             0.1.3.1
```
Comment (domenkozar) [inline stack2nix.cabal]: I think we should bump to 0.2 once we fix the last outstanding bug :)

Comment (jmitchell) [reply stack2nix.cabal]: 👍

File: src/Stack2nix/External/Stack.hs [hunk 5]
```diff
@@ -142,10 +146,11 @@ planAndGenerate boptsCli baseDir outDir remoteUri revPkgs argRev threads doAfter
...
+  void $ liftIO $ mapM_ (\pkg -> cabal2nix ("cabal://" ++ pkg) Nothing Nothing (Just outDir) hackageDB) $ words "hscolour stringbuilder"
```
Comment (jmitchell) [inline src/Stack2nix/External/Stack.hs]: I want to kill this hack. Some packages will build if more hardcoded packages are added here, but ideally there shouldn't be any.

File: src/Stack2nix/External/Stack.hs (hunk 3 above)
Comment (jmitchell) [reply src/Stack2nix/External/Stack.hs]: We worked out that the `hSnapshot` is the timestamp being passed to `loadHackageDB`, and that subsequent calls are reusing the resulting `hackageDB`. `cabal2nix` has the role of working out the DB's path, hence `Nothing`.

**Comment by domenkozar** [comment]
@jmitchell I think if we pin hackage-db while checking stack2nix output we're good to merge this as a stop-gap.

**Pull request merged**